ops/pebble.py

```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                    self._part_type = self._prepare_for_part_body()
+
+                    # Advance the buffer to point at the data.
+                    self._buffer = self._buffer[data_end:]
+                elif self._buffer.startswith(self._terminal_boundary):
+                    # We've reached the terminal boundary; ignore any remaining data
+                    self.terminated = True
+                    break
+                else:
+                    # Insufficient data to continue
+                    break
+
+            if self._part_type == 'response':
+                next_boundary_index = self._get_next_boundary_index()
+                if next_boundary_index is not None:
+                    data = self._buffer[:next_boundary_index]
+                    self._response = json.loads(data.decode())
+                    # Advance the buffer to point at the next boundary
+                    self._buffer = self._buffer[next_boundary_index:]
+                    self._headers = None
+                else:
+                    # Insufficient data to continue
+                    break
+            elif self._part_type == 'files':
+                next_boundary_index = self._get_next_boundary_index()
+                if next_boundary_index is not None:
+                    # Next boundary's location is known; this is the final write.
+                    data = self._buffer[:next_boundary_index]
+                    outfile = self._get_open_tempfile()
+                    outfile.write(data)
+                    outfile.close()
+
+                    # Advance the buffer to point at the next boundary
+                    self._buffer = self._buffer[next_boundary_index:]
+                    self._headers = None
+                else:
+                    # Data present but boundary not yet found.
+                    # Write data from buffer, but don't accidentally write data which may belong
+                    # to the next boundary line.
+                    safe_bound = max(0, len(self._buffer) - len(self._terminal_boundary))
+                    data = self._buffer[:safe_bound]
+
+                    # Don't open/close file for intermediate writes
+                    outfile = self._get_open_tempfile()
+                    outfile.write(data)
+
+                    # Advance the buffer for bytes consumed
+                    self._buffer = self._buffer[safe_bound:]
+
+                    # Insufficient data to continue
+                    break
+
+    def _parse_headers(self, data):
+        parser = email.parser.BytesFeedParser()
+        parser.feed(data)
+        message = parser.close()
+        return message
+
+    def _detect_invalid_content_disposition(self):
+        content_disposition = self._headers.get_content_disposition()
+        if content_disposition != 'form-data':
+            raise RuntimeError(
+                'unexpected content disposition: {}'.format(content_disposition))
```

Comment:
Similar to earlier comment, let's use the repr when dealing with bad data as we don't know what kind of garbage made it invalid.

Also, this should not be a RuntimeError. We're parsing some data from the network, and got a value we don't expect. This isn't an error in the local runtime.

I believe we're using ProtocolError for that kind of issue in this file, but please have a look.  There are other cases like this that need fixing too. Would you mind to have a look please?

Reply:
Switched all runtime errors to protocol errors with repr.